File: src/protocol/jsonBuilder.ts
/**
 * TS.43 JSON Response Builder
 *
 * Converts the internal ServiceEntitlementResponse into the GSMA TS.43
 * JSON format with Vers, Token, and per-application blocks.
 */

import type { ServiceEntitlementResponse, ApplicationConfig } from './responseTypes.js';

/**
 * Build a TS.43 JSON response object.
 *
 * Output format:
 * {
 *   "Vers": { "version": "1", "validity": "172800" },
 *   "Token": { "token": "..." },
 *   "ap2004": { "EntitlementStatus": "1", ... }
 * }
 */
export function buildJsonResponse(response: ServiceEntitlementResponse): object {
  const result: Record<string, unknown> = {
    Vers: {
      version: response.version,
      validity: String(response.validity),
    },
    Token: {
      token: response.token,
    },
  };

  for (const app of response.applications) {
    result[app.appId] = buildApplicationBlock(app);
  }

  return result;
}

function buildApplicationBlock(app: ApplicationConfig): Record<string, unknown> {
  const block: Record<string, unknown> = {
    EntitlementStatus: String(app.entitlementStatus),
  };

  if (app.addrStatus !== undefined) {
    block.AddrStatus = String(app.addrStatus);
  }

  if (app.tcStatus !== undefined) {
    block.TC_Status = String(app.tcStatus);
  }

  if (app.provStatus !== undefined) {
    block.ProvStatus = String(app.provStatus);
  }

  if (app.serviceFlowUrl) {
    block.ServiceFlow_URL = app.serviceFlowUrl;
  }

  if (app.addresses && app.addresses.length > 0) {
    const addrBlock: Record<string, { AddrType: string; Addr: string }> = {};
    for (let i = 0; i < app.addresses.length; i++) {
      const addr = app.addresses[i]!;
      addrBlock[String(i + 1)] = {
        AddrType: addr.addrType,
        Addr: addr.addr,
      };
    }
    block.Addr = addrBlock;
  }

  if (app.extraParams) {
    for (const [key, value] of Object.entries(app.extraParams)) {
      block[key] = value;
    }
  }

  return block;
}

File: internal/protocol/json_builder.go
package protocol

import "strconv"

// BuildJSONResponse converts a ServiceEntitlementResponse to the GSMA TS.43 JSON format.
//
// Output format:
//
//	{
//	  "Vers": { "version": "1", "validity": "172800" },
//	  "Token": { "token": "..." },
//	  "ap2004": { "EntitlementStatus": "1", ... }
//	}
func BuildJSONResponse(response *ServiceEntitlementResponse) map[string]interface{} {
	result := map[string]interface{}{
		"Vers": map[string]string{
			"version":  response.Version,
			"validity": strconv.Itoa(response.Validity),
		},
		"Token": map[string]string{
			"token": response.Token,
		},
	}

	for _, app := range response.Applications {
		result[app.AppID] = buildApplicationBlock(&app)
	}

	return result
}

func buildApplicationBlock(app *ApplicationConfig) map[string]interface{} {
	block := map[string]interface{}{
		"EntitlementStatus": strconv.Itoa(app.EntitlementStatus),
	}

	if app.AddrStatus != nil {
		block["AddrStatus"] = strconv.Itoa(*app.AddrStatus)
	}

	if app.TcStatus != nil {
		block["TC_Status"] = strconv.Itoa(*app.TcStatus)
	}

	if app.ProvStatus != nil {
		block["ProvStatus"] = strconv.Itoa(*app.ProvStatus)
	}

	if app.ServiceFlowURL != "" {
		block["ServiceFlow_URL"] = app.ServiceFlowURL
	}

	if len(app.Addresses) > 0 {
		addrBlock := map[string]interface{}{}
		for i, addr := range app.Addresses {
			addrBlock[strconv.Itoa(i+1)] = map[string]string{
				"AddrType": addr.AddrType,
				"Addr":     addr.Addr,
			}
		}
		block["Addr"] = addrBlock
	}

	for k, v := range app.ExtraParams {
		block[k] = v
	}

	return block
}
